File: testbed_ruby/test/associations/m_n_to_m_n_test.rb
require 'test/unit'
require './src-gen-umple/student_s'
require './src-gen-umple/mentor_s'
require './src-gen-umple/program_s'

module CruiseAssociations
class MNToMNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentS.new(99)
    assert_equal(0,s.number_of_mentors)
  end

  def test_CreateMentorWithoutStudent
    m = MentorS.new("blah")
    assert_equal(0,m.number_of_students)
  end

  def test_SetStudentsJustEnough
    m = MentorS.new("blah")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    
    assert_equal(true,m.set_students([s,s2,s3]))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
  end  
  
  def test_SetStudentsAtMax
    m = MentorS.new("blah")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    
    assert_equal(true,m.set_students([s,s2,s3,s4,s5]))
    
    assert_equal(5,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))
    assert_equal(m,s5.get_mentor(0))
  end   
  
  def test_AddStudents
    m = MentorS.new("blah")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    s6 = StudentS.new(94)

    assert_equal(true,m.add_student(s))
    assert_equal(true,m.add_student(s2))
    assert_equal(true,m.add_student(s3))
    assert_equal(true,m.add_student(s4))
    assert_equal(true,m.add_student(s5))
    assert_equal(false,m.add_student(s6))
    
    assert_equal(5,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))
    assert_equal(m,s5.get_mentor(0))
    assert_equal(0,s6.number_of_mentors)
  end   

  def test_RemoveMiddleStudentWhenNotValidMaintainsTheOrder
    m = MentorS.new("blah")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)


    assert_equal(true,m.set_students([s,s2,s3]))
    assert_equal(false,m.remove_student(s2))
    
    assert_equal(3,m.number_of_students)
    assert_equal(s2,m.get_student(1))
  end 

  def test_RemoveStudents
    m = MentorS.new("blah")
    m2 = MentorS.new("blah2")
    m3 = MentorS.new("blah3")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    s6 = StudentS.new(94)

    m.set_students([s,s2,s3,s4,s5])
    m2.set_students([s,s2,s3,s4,s5])
    m3.set_students([s,s2,s3,s4,s5])
    
    
    assert_equal(false,m.remove_student(s6))
    assert_equal(true,m.remove_student(s5))
    
    assert_equal(2,s5.number_of_mentors)
    assert_equal(4,m.number_of_students)
  end 

  def test_SetStudentsTooManyAndTooFew
    m = MentorS.new("blah")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    s6 = StudentS.new(94)
    
    assert_equal(true,m.set_students([s,s2,s3]))
    
    assert_equal(false,m.set_students([s4,s5]))
    assert_equal(false,m.set_students([s4,s5,s,s2,s3,s6]))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
    assert_equal(0,s6.number_of_mentors)
  end   

  def test_MentorAlreadyHasEnoughStudents
    m = MentorS.new("blah")
    
    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    
    m.set_students([s,s2,s3,s4,s5])
    
    s6 = StudentS.new(94)
    assert_equal(false,m.add_student(s6))
    
    assert_equal(5,m.number_of_students)
    assert_equal(0,s6.number_of_mentors)
  end    

  def test_replaceMentor
    m = MentorS.new("blah")
    m2 = MentorS.new("blah2")
    m3 = MentorS.new("blah3")

    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    
    m.set_students([s,s2,s3,s4,s5])
    assert_equal(true,s.set_mentors([m2,m3]))

    assert_equal(4,m.number_of_students)
    assert_equal(1,m2.number_of_students)
    assert_equal(1,m3.number_of_students)
    assert_equal(2,s.number_of_mentors)
  end 

  def test_is_number_of_students_valid
    m = MentorS.new("blah")
    assert_equal(false,m.is_number_of_students_valid)

    m.add_student(StudentS.new(99))
    assert_equal(false,m.is_number_of_students_valid)

    m.add_student(StudentS.new(98))
    assert_equal(false,m.is_number_of_students_valid)

    m.add_student(StudentS.new(97))
    assert_equal(true,m.is_number_of_students_valid)

  end

  def test_getBoundsForStudent
    assert_equal(3,MentorS::minimum_number_of_students)
    assert_equal(5,MentorS::maximum_number_of_students)
  end  

  def test_deleteMentorAndStudentHasEnough
    m = MentorS.new("blah")
    m2 = MentorS.new("blah2")
    m3 = MentorS.new("blah3")

    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    
    m.set_students([s,s2,s3,s4,s5])
    m2.set_students([s,s2,s3,s4,s5])
    m3.set_students([s,s2,s3,s4,s5])
    
    p = ProgramS.new
    s.set_program(p)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(p,s.get_program)
    assert_equal(2,s.number_of_mentors)
  end  

  def test_deleteMentorAndStudentNowHasTooFewMentors
    m = MentorS.new("blah")
    m2 = MentorS.new("blah2")

    s = StudentS.new(99)
    s2 = StudentS.new(98)
    s3 = StudentS.new(97)
    s4 = StudentS.new(96)
    s5 = StudentS.new(95)
    
    m.set_students([s,s2,s3,s4,s5])
    m2.set_students([s,s2,s3])
    
    p = ProgramS.new
    s.set_program(p)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    assert_equal(0,s.number_of_mentors)
  end  

end
end

File: testbed_php/test/associations/MNToMNTest.php
<?php

class MNToMNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentS(99);
    $this->assertEqual(0,$s->numberOfMentors());
  }
  

  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorS("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }
  

  public function test_SetStudentsJustEnough()
  {
    $m = new MentorS("blah");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    
    $this->assertEqual(true,$m->setStudents(array($s,$s2,$s3)));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
  }  
  

  public function test_SetStudentsAtMax()
  {
    $m = new MentorS("blah");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    
    $this->assertEqual(true,$m->setStudents(array($s,$s2,$s3,$s4,$s5)));
    
    $this->assertEqual(5,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));
    $this->assertEqual($m,$s5->getMentor(0));
  }   
  

  public function test_AddStudents()
  {
    $m = new MentorS("blah");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    $s6 = new StudentS(94);

    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(true,$m->addStudent($s2));
    $this->assertEqual(true,$m->addStudent($s3));
    $this->assertEqual(true,$m->addStudent($s4));
    $this->assertEqual(true,$m->addStudent($s5));
    $this->assertEqual(false,$m->addStudent($s6));
    
    $this->assertEqual(5,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));
    $this->assertEqual($m,$s5->getMentor(0));
    $this->assertEqual(0,$s6->numberOfMentors());
  }   
  

  public function test_RemoveMiddleStudentWhenNotValidMaintainsTheOrder()
  {
    $m = new MentorS("blah");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);


    $this->assertEqual(true,$m->setStudents(array($s,$s2,$s3)));
    $this->assertEqual(false,$m->removeStudent($s2));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($s2,$m->getStudent(1));
  } 
  

  public function test_RemoveStudents()
  {
    $m = new MentorS("blah");
    $m2 = new MentorS("blah2");
    $m3 = new MentorS("blah3");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    $s6 = new StudentS(94);

    $m->setStudents(array($s,$s2,$s3,$s4,$s5));
    $m2->setStudents(array($s,$s2,$s3,$s4,$s5));
    $m3->setStudents(array($s,$s2,$s3,$s4,$s5));
    
    
    $this->assertEqual(false,$m->removeStudent($s6));
    $this->assertEqual(true,$m->removeStudent($s5));
    
    $this->assertEqual(2,$s5->numberOfMentors());
    $this->assertEqual(4,$m->numberOfStudents());
  } 
  

  public function test_SetStudentsTooManyAndTooFew()
  {
    $m = new MentorS("blah");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    $s6 = new StudentS(94);
    
    $this->assertEqual(true,$m->setStudents(array($s,$s2,$s3)));
    
    $this->assertEqual(false,$m->setStudents(array($s4,$s5)));
    $this->assertEqual(false,$m->setStudents(array($s4,$s5,$s,$s2,$s3,$s6)));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
    $this->assertEqual(0,$s6->numberOfMentors());
  }   
  

  public function test_MentorAlreadyHasEnoughStudents()
  {
    $m = new MentorS("blah");
    
    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    
    $m->setStudents(array($s,$s2,$s3,$s4,$s5));
    
    $s6 = new StudentS(94);
    $this->assertEqual(false,$m->addStudent($s6));
    
    $this->assertEqual(5,$m->numberOfStudents());
    $this->assertEqual(0,$s6->numberOfMentors());
  }    


  public function test_replaceMentor()
  {
    $m = new MentorS("blah");
    $m2 = new MentorS("blah2");
    $m3 = new MentorS("blah3");

    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    
    $m->setStudents(array($s,$s2,$s3,$s4,$s5));
    $this->assertEqual(true,$s->setMentors(array($m2,$m3)));

    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(1,$m2->numberOfStudents());
    $this->assertEqual(1,$m3->numberOfStudents());
    $this->assertEqual(2,$s->numberOfMentors());
  } 

  

  public function test_isNumberOfStudentsValid()
  {
    $m = new MentorS("blah");
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    $m->addStudent(new StudentS(99));
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    $m->addStudent(new StudentS(98));
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    $m->addStudent(new StudentS(97));
    $this->assertEqual(true,$m->isNumberOfStudentsValid());

  }
  

  public function test_getBoundsForStudent()
  {
    $this->assertEqual(3,MentorS::minimumNumberOfStudents());
    $this->assertEqual(5,MentorS::maximumNumberOfStudents());
  }  
  

  public function test_deleteMentorAndStudentHasEnough()
  {
    
    $m = new MentorS("blah");
    $m2 = new MentorS("blah2");
    $m3 = new MentorS("blah3");

    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    
    $m->setStudents(array($s,$s2,$s3,$s4,$s5));
    $m2->setStudents(array($s,$s2,$s3,$s4,$s5));
    $m3->setStudents(array($s,$s2,$s3,$s4,$s5));
    
    $p = new ProgramS();
    $s->setProgram($p);
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual($p,$s->getProgram());
    $this->assertEqual(2,$s->numberOfMentors());
  }  


  public function test_deleteMentorAndStudentNowHasTooFewMentors()
  {
    
    $m = new MentorS("blah");
    $m2 = new MentorS("blah2");

    $s = new StudentS(99);
    $s2 = new StudentS(98);
    $s3 = new StudentS(97);
    $s4 = new StudentS(96);
    $s5 = new StudentS(95);
    
    $m->setStudents($s,$s2,$s3,$s4,$s5);
    $m2->setStudents($s,$s2,$s3);
    
    $p = new ProgramS();
    $s->setProgram($p);
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());
  }  

}
?>